include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,312 @@
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bin_edges The bin-edges.
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        E1&& data, E2&& bin_edges, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        return histogram(data, bin_edges, xt::ones<value_type>({data.size()}), density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bins The number of bins. [default: 10]
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1>
+    inline auto histogram(
+        E1&& data, size_t bins=10, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        auto bin_edges = histogram_bin_edges(data, xt::ones<value_type>({data.size()}), bins);
+
+        return histogram(data, bin_edges, xt::ones<value_type>({data.size()}), density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bins The number of bins.
+     * @param weights Weight factors corresponding to each data-point.
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        E1&& data, size_t bins, E2&& weights, bool density=false
+    )
+    {
+        auto bin_edges = histogram_bin_edges(data, weights, bins);
+
+        return histogram(data, bin_edges, weights, density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Defines different algorithms in "histogram_bin_edges"
+     */
+    enum class histogram_algorithm { automatic, linspace, uniform };
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the bin-edges of a histogram of a set of data using different algorithms.
+     *
+     * @param data The data.
+     * @param weights Weight factors corresponding to each data-point.
+     * @param left The lower-most edge.
+     * @param right The upper-most edge.
+     * @param bins The number of bins. [default: 10]
+     * @param mode The type of algorithm to use. [default: "auto"]
+     * @return An one-dimensional xarray<double>, length: bins+1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram_bin_edges(const E1 &data, const E2 &weights,
+        E3 left, E3 right, size_t bins=10, histogram_algorithm mode=histogram_algorithm::automatic
```

Comment:
Universal references should be used here.

Reply:
@JohanMabille This is not a wrapper function. I.e. should not use universal references?

Reply:
I think you can use universal references whatever the type of the function is, it is not related to being a wrapper function or not.